geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java

```diff
@@ -764,24 +774,28 @@ public void endGiiQueueing() {
             }
             if (value instanceof HAEventWrapper) {
               if (((HAEventWrapper) value).getClientUpdateMessage() == null) {
-                // if there is no wrapped message look for it in the HA container map
-                ClientUpdateMessageImpl haContainerMessage =
-                    (ClientUpdateMessageImpl) haContainer.get(value);
-                if (haContainerMessage != null) {
-                  ((HAEventWrapper) value).setClientUpdateMessage(haContainerMessage);
-                } else {
-                  if (isDebugEnabled) {
-                    logger.debug(
-                        "{} ATTENTION: found gii queued event with null event message.  Please see bug #44852: {}",
-                        this.regionName, value);
-                  }
-                  continue;
+                if (isDebugEnabled) {
+                  logger.debug(
+                      "{} ATTENTION: found gii queued event with null event message.  Please see bug #44852: {}",
+                      this.regionName, value);
                 }
+                continue;
               }
             }
+
             basicPut(value);
+
+            // The HAEventWrapper ref count must be decremented because it was
+            // incremented when it was queued in giiQueue.
             if (value instanceof HAEventWrapper) {
-              decAndRemoveFromHAContainer((HAEventWrapper) value);
```

Comment:
Original code did a decAndRemove here ... are we accidentally leaving in HAContainer?
Should we use checkAndRemoveFromClientMsgsRegion here?

Reply:
We determined that we could leak the event so we wrote a test and changed this code to prevent that from happening by invoking CacheClientNotifier.checkAndRemoveFromClientMsgsRegion().  This will both decrement the PutInProgressCount and remove from the container if necessary.